build/lib/src/library_cycle_graph/library_cycle_graph_loader.dart

```diff
@@ -340,20 +408,36 @@ class LibraryCycleGraphLoader {
   ///
   /// Previously computed state is used if possible, anything additional is
   /// loaded using [assetDepsLoader].
+  ///
+  /// See class note about recursive calls.
   Future<PhasedValue<LibraryCycle>> libraryCycleOf(
     AssetDepsLoader assetDepsLoader,
     AssetId id,
   ) async {
-    await _load(assetDepsLoader, id);
-    _buildCycles(assetDepsLoader.phase);
-    return _cycles[id]!;
+    if (_runningAtPhases.isNotEmpty &&
+        assetDepsLoader.phase >= _runningAtPhases.last) {
+      throw StateError(
+        'Cannot recurse at later or equal phase ${assetDepsLoader.phase}, '
+        'already running at : $_runningAtPhases',
+      );
+    }
+    _runningAtPhases.add(assetDepsLoader.phase);
+    try {
+      await _load(assetDepsLoader, id);
+      _buildCycles(assetDepsLoader.phase);
+      return _cycles[id]!;
+    } finally {
+      _runningAtPhases.removeLast();
```

Comment:
I assume this is assumed to be the same as the one added above..? Maybe assert that?

Reply:
Good idea, thanks, done.

Removed the try/finally since the linter doesn't like me throwing inside a "finally", the code isn't actually designed to work when there are exceptions anyway.